File: tests/metrics/parametric_expected_shortfall_test.py
import numpy as np
import pytest
from scipy.stats import norm
from taflow.metrics import ParametricExpectedShortfall

@pytest.mark.parametrize(('values', 'cutoff'), [(np.array([-0.02, 0.01, 0.03, -0.01, 0.015]), 0.05), (np.linspace(-0.04, 0.05, 101), 0.01), (np.random.default_rng(20260812).normal(0.0004, 0.012, 513), 0.1), (np.array([np.nan, 0.02, -0.03, np.nan, 0.01]), 0.025)])
def test_matches_scipy_gaussian_expected_shortfall(values, cutoff):
    usable = values[~np.isnan(values)]
    expected = float(usable.mean() - usable.std(ddof=1) * norm.pdf(norm.ppf(cutoff)) / cutoff)
    actual = ParametricExpectedShortfall(cutoff=cutoff).from_returns(values).compute()
    assert actual == pytest.approx(expected, rel=2e-08, abs=2e-10)

def test_input_methods_and_streaming_lifecycle():
    returns = np.array([0.01, -0.02, 0.03, -0.015])
    equity = 100.0 * np.cumprod(np.r_[1.0, 1.0 + returns])
    expected = ParametricExpectedShortfall().from_returns(returns).compute()
    assert ParametricExpectedShortfall().from_log_returns(np.log1p(returns)).compute() == pytest.approx(expected)
    assert ParametricExpectedShortfall().from_equity(equity).compute() == pytest.approx(expected)
    assert ParametricExpectedShortfall().from_pnl(np.diff(equity), initial_capital=100.0).compute() == pytest.approx(expected)
    state = ParametricExpectedShortfall().from_returns([])
    assert state.append(returns[0]).compute() is None
    assert state.extend(returns[1:]).compute() == pytest.approx(expected)
    assert state.reset().extend(returns).compute() == pytest.approx(expected)

def test_validation_and_constant_edge():
    unbound = ParametricExpectedShortfall()
    with pytest.raises(ValueError):
        unbound.append(0.01)
    assert ParametricExpectedShortfall().from_returns([0.01]).compute() is None
    assert ParametricExpectedShortfall().from_returns([0.01, 0.01]).compute() == pytest.approx(0.01)
    with pytest.raises(ValueError):
        ParametricExpectedShortfall(cutoff=0.0).from_returns([0.01, -0.01])
    with pytest.raises(ValueError):
        ParametricExpectedShortfall().from_returns([np.inf])

File: crates/taflow-metrics/src/metrics/parametric_expected_shortfall_test.rs
use super::ParametricExpectedShortfall;
use crate::{MetricInputKind, NanPolicy};
#[test]
fn computes_gaussian_lower_tail_mean() {
    let mut m = ParametricExpectedShortfall::new(0.05, NanPolicy::Omit)
        .and_then(|mut state| {
            state.from_returns(&[])?;
            Ok(state)
        })
        .unwrap();
    let actual = m
        .extend(&[-0.02, 0.01, 0.03, -0.01, 0.015])
        .unwrap()
        .unwrap();
    assert!((actual - (-0.036_254_256_150_148_51)).abs() < 2e-10);
}
#[test]
fn input_methods_lifecycle_and_missing_are_invariant() {
    let returns: [f64; 3] = [0.01, -0.02, 0.03];
    let logs: Vec<_> = returns.iter().map(|v| v.ln_1p()).collect();
    let equity = [100., 101., 98.98, 101.9494];
    let pnl = [1., -2.02, 2.9694];
    let mut a = ParametricExpectedShortfall::new(0.05, NanPolicy::Omit)
        .and_then(|mut state| {
            state.from_returns(&[])?;
            Ok(state)
        })
        .unwrap();
    let mut b = ParametricExpectedShortfall::new(0.05, NanPolicy::Omit)
        .and_then(|mut state| {
            state.from_log_returns(&[])?;
            Ok(state)
        })
        .unwrap();
    let mut c = ParametricExpectedShortfall::new(0.05, NanPolicy::Omit)
        .and_then(|mut state| {
            state.from_equity(&[])?;
            Ok(state)
        })
        .unwrap();
    let mut d = ParametricExpectedShortfall::new(0.05, NanPolicy::Omit)
        .and_then(|mut state| {
            state.from_pnl(&[], 100.)?;
            Ok(state)
        })
        .unwrap();
    let expected = a.extend(&returns).unwrap().unwrap();
    assert!((b.extend(&logs).unwrap().unwrap() - expected).abs() < 1e-14);
    assert!((c.extend(&equity).unwrap().unwrap() - expected).abs() < 1e-14);
    assert!((d.extend(&pnl).unwrap().unwrap() - expected).abs() < 1e-14);
    a.reset();
    assert_eq!(
        a.extend(&[returns[0], f64::NAN, returns[1], returns[2]])
            .unwrap(),
        Some(expected)
    );
}
#[test]
fn validates_cutoff_domain_and_warmup() {
    assert!(ParametricExpectedShortfall::new(0., NanPolicy::Omit)
        .and_then(|mut state| {
            state.from_returns(&[])?;
            Ok(state)
        })
        .is_err());
    assert!(ParametricExpectedShortfall::new(0.05, NanPolicy::Omit)
        .and_then(|mut state| {
            state.append(0.0)?;
            Ok(state)
        })
        .is_err());
    let mut m = ParametricExpectedShortfall::new(0.05, NanPolicy::Omit)
        .and_then(|mut state| {
            state.from_returns(&[])?;
            Ok(state)
        })
        .unwrap();
    assert_eq!(m.append(0.01).unwrap(), None);
    assert_eq!(m.append(0.01).unwrap(), Some(0.01));
}
